### PR TITLE
Fix bugs and Add affiliated feature.

### PR DESCRIPTION
**Pull request by leexx227**

### CODE REVIEW

File: HighAvailabilityModule.Algorithm/MembershipWithWitness.cs [hunk 1]
```diff
@@ -31,7 +31,11 @@ public class MembershipWithWitness
 
         internal CancellationToken AlgorithmCancellationToken => this.AlgorithmCancellationTokenSource.Token;
 
-        public MembershipWithWitness(IMembershipClient client, TimeSpan heartBeatInterval, TimeSpan heartBeatTimeout)
+        private string AffiliatedType;
```
Comment (amat27) [inline HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: Better use the word affinity

File: HighAvailabilityModule.Algorithm/MembershipWithWitness.cs [hunk 2]
```diff
@@ -143,32 +147,66 @@ internal async Task HeartBeatAsPrimaryAsync()
...
+        {
+            var affiliatedPrimary = this.AffiliatedType == string.Empty || (this.AffiliatedPrimaryUp && this.lastSeenAffiliated.Entry.Uname.ToLower() == this.Uname.ToLower() && now - this.lastSeenAffiliated.QueryTime < (this.HeartBeatTimeout - this.HeartBeatInterval));
+            if (!affiliatedPrimary)
+            {
+                Trace.TraceInformation($"[{DateTime.UtcNow:O}][Protocol] Affiliated service is running on another machine. {this.Dump()}.");
+            }
+            return affiliatedPrimary;
+        }
+
+        internal async Task CheckAffiliatedAsync(DateTime now)
```
Comment (amat27) [inline HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: This is basically the same as `HeartBeatAsPrimaryAsync`? Can we consolidate the logic?

File: HighAvailabilityModule.Algorithm/MembershipWithWitness.cs [hunk 3]
```diff
@@ -143,32 +147,66 @@ internal async Task HeartBeatAsPrimaryAsync()
...
             var primary = this.PrimaryUp && this.lastSeenHeartBeat.Entry.Uuid == this.Uuid && now - this.lastSeenHeartBeat.QueryTime < (this.HeartBeatTimeout - this.HeartBeatInterval);
             if (!primary)
             {
                 Trace.TraceInformation($"[{DateTime.UtcNow:O}][Protocol] Running as secondary {this.Dump()}.");
             }
-
             return primary;
         }
 
+        internal bool AffiliatedASPrimary(DateTime now)
```
Comment (amat27) [inline HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: This is basically the same as `CheckPrimaryAsync`?
Also, the name needs to apply camel case

File: HighAvailabilityModule.UnitTest/AlgorithmTest.cs [hunk 4]
```diff
@@ -94,43 +97,96 @@ public async Task RunningAsPrimaryTest4()
...
+            Assert.IsFalse(this.algo3.RunningAsPrimary(now));
+        }
+
+        [TestMethod]
+        public async Task AffiliatedAsPrimary1()
+        {
+            DateTime now = DateTime.UtcNow;
+            this.server.CurrentTable = new Dictionary<string, HeartBeatEntry>();
+            await this.algo.CheckAffiliatedAsync(now);
+            Assert.IsTrue(this.algo.AffiliatedASPrimary(now));
```
Comment (amat27) [inline HighAvailabilityModule.UnitTest/AlgorithmTest.cs]: Can we also check `algo.RunAsPrimay` at the end?

File: keys/files.txt [hunk 5]
```diff
@@ -0,0 +1,300 @@
+################################################
```
Comment (amat27) [inline keys/files.txt]: We don't nee this?

**Review by amat27 — CHANGES_REQUESTED** [review]

File: HighAvailabilityModule.Algorithm/MembershipWithWitness.cs [hunk 6]
```diff
@@ -23,19 +24,22 @@ public class MembershipWithWitness
...
         private CancellationTokenSource AlgorithmCancellationTokenSource { get; } = new CancellationTokenSource();
 
         internal CancellationToken AlgorithmCancellationToken => this.AlgorithmCancellationTokenSource.Token;
 
-        private string AffiliatedType;
+        private string AffinityType;
+
+        private const string LastSeenHeartBeatString = "LastSeenHeartBeat";
+
+        private const string lastSeenAffinityString = "lastSeenAffinity";
```
Comment (amat27) [inline HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: Upper case initial

File: HighAvailabilityModule.UnitTest/AlgorithmTest.cs [hunk 7]
```diff
@@ -135,51 +135,59 @@ public async Task RunningAsPrimaryTest8()
         }
 
         [TestMethod]
-        public async Task AffiliatedAsPrimary1()
+        public async Task AffinityAsPrimary1()
```
Comment (amat27) [inline HighAvailabilityModule.UnitTest/AlgorithmTest.cs]: We need test cases for `RunAsPrimary` with affinity

File: HighAvailabilityModule.Algorithm/MembershipWithWitness.cs [hunk 8]
```diff
@@ -77,13 +86,14 @@ public void Stop()
...
             {
                 token.ThrowIfCancellationRequested();
                 await Task.Delay(this.HeartBeatInterval, token);
                 await this.CheckPrimaryAsync(DateTime.UtcNow).ConfigureAwait(false);
+                await this.CheckAffinityAsync(DateTime.UtcNow).ConfigureAwait(false);
 
-                if (!this.PrimaryUp)
+                if (!this.PrimaryUp && this.AffinityAsPrimary(DateTime.UtcNow))
                 {
                     Trace.TraceWarning($"[{DateTime.UtcNow:O}][Protocol][{this.Uuid}] Primary down");
```
Comment (amat27) [inline HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: Also change the `Trace` to `TraceSource`

**Review by amat27 — CHANGES_REQUESTED** [review]

File: HighAvailabilityModule.Storage.Server/startup.sql [hunk 9]
```diff
@@ -38,9 +38,9 @@ IF OBJECT_ID('SetDataEntry') IS NOT NULL
 	DROP PROCEDURE SetDataEntry;
 GO
 CREATE PROCEDURE SetDataEntry
-	@dpath nvarchar(50),
-	@dkey nvarchar(50),
-	@dvalue nvarchar(50),
+	@dpath nvarchar(4000),
```
Comment (amat27) [inline HighAvailabilityModule.Storage.Server/startup.sql]: Can we use [nvarchar(max)](https://stackoverflow.com/questions/11131958/what-is-the-maximum-characters-for-the-nvarcharmax/11131977) ?

Comment (leexx227) [reply HighAvailabilityModule.Storage.Server/startup.sql]: Primary keys cannot set to be `max` and the max length is 900.

File: HighAvailabilityModule.Algorithm/MembershipWithWitness.cs [hunk 10]
```diff
@@ -30,7 +30,7 @@ public class MembershipWithWitness
 
         internal CancellationToken AlgorithmCancellationToken => this.AlgorithmCancellationTokenSource.Token;
 
-        private static TraceSource ts = new TraceSource("Microsoft.Hpc.Highavailablity");
+        public static TraceSource ts = new TraceSource("HpcHighAvailablity.Algorithm");
```
Comment (amat27) [inline HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: Why do we remove the `Microsoft.Hpc` prefix?

Comment (amat27) [reply HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: `public static readonly`

Comment (amat27) [reply HighAvailabilityModule.Algorithm/MembershipWithWitness.cs]: Let's aslo add the sample code which listens the trace source

**Review by amat27 — CHANGES_REQUESTED** [review]

**Pull request merged**